### ISSUE TITLE
Async support

### ISSUE THREAD

**Issue by damusix** [hapijs/vision#177]
It would be nice to support async context in view manager.

My use case is to query cached data from catbox client and pass it to frontend to generate menus and lists, etc.

EG: 
```js

server.views({
    engines: {
        html: Handlebars
    },
    ...
    context: async () => {

        return {
            posts: await server.methods.getCachedPosts()
        }
    }
});
```

Currently, I have to hoist the response into an object store and wait for it to populate before it reflects in my frontend, which means I need to refresh that object store whenever cache expires.

**Comment by wswoodruff** [comment]
I'm down for that! Thanks for pointing this out — I'd like all the places you find something like `=== 'function'` to support async/await in the manager. Lemme know if you're interested in making a PR for that

**Comment by damusix** [comment]
Awesome! I'll start something this week. Seems like a couple things that need to become async/await.

**Comment by damusix** [comment]
`promisify` did the trick!

**Comment by wswoodruff** [comment]
Oh nice! That's a good work around til we get this fix in

**Comment by hueniverse** [comment]
I didn't yet. The commit message has the wrong issues...

**Comment by wswoodruff** [comment]
Ohh gotcha — I thought you reopened so we could finish up the discussion. Ok that's fine! Are you planning on implementing it or should we restart off your work?

**Comment by hueniverse** [comment]
I will probably do a breaking release in the near future with support for async methods all around.

**Comment by wswoodruff** [comment]
Ok good to know, and thank you sounds great! Please update me if you decide you want to offload that here

**Comment by damusix** [comment]
Sounds good gentlemen. Looking forward to it. @hueniverse if there's anything I can help with, feel free to ask. I'm in the `hapi hour` slack channel.

**Comment by lloydbenson** [comment]
Hi @damusix and/or @wswoodruff. I think this conversation got dropped. I think we'd be open to a PR if one of you wants to look into this!

**Comment by wswoodruff** [comment]
@lloydbenson Yeah good call this is unblocked since Eran won't be working on it — this task will take significant effort, not sure when I'll be able to contribute that amount of time but I'll keep this in mind

**Comment by lloydbenson** [comment]
If neither of you can do it, that is fine we can just close it up and do a new issue later if someone does have time in the future.

**Comment by damusix** [comment]
I had a PR on this a while back that I can restore at some point @lloydbenson @wswoodruff https://github.com/hapijs/vision/pull/179

**Comment by lloydbenson** [comment]
nice! Thanks. Before you do that work I think that gives the community a good indication of what you did and lets try to get some input from the community on what they thought of that solution (I don't use vision that much so better to get others feedback). I mostly just wanted to make sure we came to a conclusion either way ;) I could be reading it wrong, but it seems like this may be a breaking change, that would make sense for a new release but feel free to correct me if I am mistaken.

**Comment by damusix** [comment]
@lloydbenson IIRC, it wasn't break any existing functionality, it was only adding the feature of being able to asynchronously render context. To your point, yes, I will wait for feedback. As Eran had stated previously, it likely warrants quite the refactor in order to properly implement async support instead of using `util.promisify` builtins, which is more of a patch.

**Comment by lloydbenson** [comment]
Thanks for the explanation! Hopefully we get some good community feedback on the changes.

**Comment by wswoodruff** [comment]
This issue was originally just about async render context support, but was expanded to planning out top-to-bottom async updates by Eran — so we can rescope this, which if we go back to just render context support would be much less effort I think, and would probably not be a breaking change.

The full async update across all vision files will be a breaking change though.

**Comment by damusix** [comment]
@wswoodruff I think it is more appropriate to rescope. I am using my a forked version in my projects anyway, so not a big deal.

**Comment by damusix** [comment]
On second thought, after reading Hapi and Vision docs more thoroughly, I think this can completely be resolved by overwriting the original decorator. Recently was able to get this to work in the following manner:

```js

// Overwriting method receives the original decorator function
const method = function (view) {

    // Return the function you want to override with
    return async function (page, context = {}, options = {}) {

        // Get the global context of required db data, etc
        const globalContext = await server.methods.globalContext();

        // Call the original function with extended context, including the page
        return view.bind(this)(
            page,
            {
                ...globalContext,
                ...context,
                page
            },
            options
        );
    };
};

server.decorate(
    'toolkit',
    'view',
    method,
    { extend: true }
);
```


New Usage:


```js

const allPrograms = {

    method: 'GET',
    path: '/',
        return await h.view('home' , {
            somestuff: await request.somestuff()
        });
}
```

**Comment by wswoodruff** [comment]
Nice — find, figuring that out, and thanks for posting some example code!

We should update `vision` itself at some point, at whichever level is worth a release but this looks like the best way to work around this so far! ☝️

I think we should keep the issue open until we get something merged we want to release —

@damusix do you think there'd be much work in re-opening your old PR for updating the render context func?

Also it'd be good to get someone from the TSC or someone with the keys to publishing this to weigh in on a release that updates just the async render context support — I'll post this in hapi hour

**Comment by damusix** [comment]
@wswoodruff OK it's up. Looks like Eran had implemented a lot of the logic I had on that PR a couple years back, so the rebase really only adds an `await context()` (which wouldn't resolve otherwise) and tests. Pretty smooth sync.

**Comment by wswoodruff** [comment]
Cool thanks for doing that @damusix!

Quicklink -> #205

**Comment by wswoodruff** [comment]
Note that async support has landed for the `context` method in `v6.1.0`

**Comment by devinivy** [comment]
I think this is all set now, but if not just give a poke!
